File: Remote/Source/Remote Elements/REView/ConnectionStatusButtonView.m
//
// ConnectionStatusButtonView.m
// Remote
//
// Created by Jason Cardwell on 5/24/11.
// Copyright 2011 Moondeer Studios. All rights reserved.
//
#import "RemoteElementView_Private.h"
#import "Button.h"
#import "ConnectionManager.h"

@implementation ConnectionStatusButtonView {}

- (void)initializeIVARs {
  [super initializeIVARs];

  self.selected = [ConnectionManager isWifiAvailable];

  [NotificationCenter
     addObserverForName:CMConnectionStatusNotification
                 object:[ConnectionManager class]
                  queue:MainQueue
             usingBlock:^(NSNotification * note) {
               if (self.model.selected != BOOLValue([note.userInfo
                                             valueForKey:CMConnectionStatusWifiAvailable]))
                 self.model.selected = !self.model.selected;
             }];
}

- (void)dealloc { [NotificationCenter removeObserver:self]; }

@end

File: UI/Source/ConnectionStatusButtonView.swift
//
//  ConnectionStatusButtonView.swift
//  Remote
//
//  Created by Jason Cardwell on 11/07/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import UIKit
import MoonKit
import DataModel
import Networking

public final class ConnectionStatusButtonView: ButtonView {

  private var receptionist: MSNotificationReceptionist!

  private var connected: Bool = false

  /** initializeIVARs */
  override func initializeIVARs() {
    super.initializeIVARs()
    connected = ConnectionManager.isWifiAvailable()
    receptionist = MSNotificationReceptionist(
      observer: self,
      forObject: nil,
      notificationName: CMConnectionStatusNotification,
      queue: NSOperationQueue.mainQueue(),
      handler: {
        (receptionist: MSNotificationReceptionist!) -> Void in
          if let v = receptionist.observer as? ConnectionStatusButtonView {
            let currentlyConnected = v.connected
            if let wifiAvailable = receptionist.notification.userInfo?[CMConnectionStatusWifiAvailableKey] as? Bool {
              if currentlyConnected != wifiAvailable { v.connected = !currentlyConnected; v.setNeedsDisplay() }
            }
          }
      })
  }

  override func drawContentInContext(ctx: CGContextRef, inRect rect: CGRect) {
    let iconColor: UIColor
    if let color = backgroundColor where color != UIColor.clearColor() { iconColor = color }
    else { iconColor = UI.DrawingKit.buttonStrokeColor }
    UI.DrawingKit.drawWifiStatus(iconColor: iconColor, connected: connected, containingFrame: rect)
  }
}